### PR TITLE
Security fix

### PR DESCRIPTION
**Pull request by id2359**
Editing or viewing a patient data (consent, demographic, clinical) record can only occur on the precondition that the user and patient are in the correct security relationship which is:

A user is the   patient administering their own data
A user is the parent or guardian of the patient
A user is a staff user ( non patient) and is in the same working  group

### CODE REVIEW

File: rdrf/rdrf/patient_view.py [hunk 1]
```diff
@@ -30,6 +30,10 @@
 from rdrf.components import RDRFContextLauncherComponent
 from rdrf.components import RDRFPatientInfoComponent
 
+from rdrf.security_checks import security_check_user_patient
+from django.core.exceptions import PermissionDenied
```
Comment (aahunter) [inline rdrf/rdrf/patient_view.py]: Minor - PermissionDenied import may be unused.

File: rdrf/rdrf/security_checks.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                                                           patient_model.pk))
+    raise PermissionDenied()
+
+def security_check_user_patient(user, patient):
+    # either user is allowed to act on this record ( return True)
+    # or not ( raise PermissionDenied error)
+    if user.is_superuser:
+        return True
+
+    if patient is None:
```
Comment (aahunter) [inline rdrf/rdrf/security_checks.py]: Can we remove this special case

File: rdrf/rdrf/security_checks.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        return True
+
+    if type(patient) is int:
+        patient_model = Patient.objects.get(id=patient_id)
+    elif type(patient) is str:
+        patient_model = Patient.objects.get(id=int(patient_id))
+    else:
+        patient_model = patient
```
Comment (aahunter) [inline rdrf/rdrf/security_checks.py]: Can we make it a pre-condition that a model is passed?

File: rdrf/rdrf/security_checks.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    except:
+        return False
+
+def _user_is_patient_type(user):
+    return any([_get_prop(user, "is_patient"),
+                _get_prop(user, "is_parent"),
+                _get_prop(user, "is_carrier")])
+
+
+def _security_violation(user, patient_model):
```
Comment (aahunter) [inline rdrf/rdrf/security_checks.py]: Call this _raise_security_violation to make the code below more obvious where the function exits.

File: rdrf/rdrf/security_checks.py (hunk 3 above)
Comment (id2359) [reply rdrf/rdrf/security_checks.py]: yes I thought it better also after thinking

File: rdrf/rdrf/security_checks.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        return True
+
+    if type(patient) is int:
+        patient_model = Patient.objects.get(id=patient_id)
+    elif type(patient) is str:
+        patient_model = Patient.objects.get(id=int(patient_id))
+    else:
+        patient_model = patient
+    
+    if _user_is_patient_type(user):
```
Comment (aahunter) [inline rdrf/rdrf/security_checks.py]: Perhaps remove on level of indentation (else can be redundant) and always raise at the end.

**Comment by aahunter** [comment]
Reviewed, subject to tests...

File: rdrf/rdrf/security_checks.py (hunk 5 above)
Comment (id2359) [reply rdrf/rdrf/security_checks.py]: nod - done

**Pull request merged**